lightning/src/ln/peers/conduit.rs

```diff
@@ -0,0 +1,191 @@
+		}
+
+		let mut messages = Vec::new();
+
+		loop {
+			// todo: find way that won't require cloning the entire buffer
+			let (current_message, offset) = self.decrypt(&read_buffer[..]);
+			if offset == 0 {
+				break;
+			}
+
+			read_buffer.drain(0..offset);
+
+			if let Some(message) = current_message {
+				messages.push(message);
+			} else {
+				break;
+			}
+		}
+
+		messages
+	}
+
+	/// Decrypt a single message. Buffer is an undelimited amount of bytes
+	pub(crate) fn decrypt(&mut self, buffer: &[u8]) -> (Option<Vec<u8>>, usize) { // the response slice should have the same lifetime as the argument. It's the slice data is read from
+		if buffer.len() < 18 {
+			return (None, 0);
+		}
+
+		let encrypted_length = &buffer[0..18]; // todo: abort if too short
+		let length_vec = chacha::decrypt(&self.receiving_key, self.receiving_nonce as u64, &[0; 0], encrypted_length).unwrap();
+		let mut length_bytes = [0u8; 2];
+		length_bytes.copy_from_slice(length_vec.as_slice());
+		let message_length = byte_utils::slice_to_be16(&length_bytes) as usize;
+
+		let message_end_index = message_length + 18 + 16; // todo: abort if too short
+		if buffer.len() < message_end_index {
+			return (None, 0);
+		}
+
+		let encrypted_message = &buffer[18..message_end_index];
+
+		self.increment_receiving_nonce();
+
+		let message = chacha::decrypt(&self.receiving_key, self.receiving_nonce as u64, &[0; 0], encrypted_message).unwrap();
+
+		self.increment_receiving_nonce();
+
+		(Some(message), message_end_index)
+	}
+
+	fn increment_sending_nonce(&mut self) {
+		Self::increment_nonce(&mut self.sending_nonce, &mut self.sending_chaining_key, &mut self.sending_key);
+	}
+
+	fn increment_receiving_nonce(&mut self) {
+		Self::increment_nonce(&mut self.receiving_nonce, &mut self.receiving_chaining_key, &mut self.receiving_key);
+	}
+
+	fn increment_nonce(nonce: &mut u32, chaining_key: &mut [u8; 32], key: &mut [u8; 32]) {
+		*nonce += 1;
+		if *nonce == 1000 {
+			Self::rotate_key(chaining_key, key);
+			*nonce = 0;
+		}
+	}
+
+	fn rotate_key(chaining_key: &mut [u8; 32], key: &mut [u8; 32]) {
+		let (new_chaining_key, new_key) = hkdf::derive(chaining_key, key);
+		chaining_key.copy_from_slice(&new_chaining_key);
+		key.copy_from_slice(&new_key);
+	}
+}
+
+#[cfg(test)]
+mod tests {
+	use hex;
+	use ln::peers::conduit::Conduit;
+
+	#[test]
+	fn test_chaining() {
+		let chaining_key_vec = hex::decode("919219dbb2920afa8db80f9a51787a840bcf111ed8d588caf9ab4be716e42b01").unwrap();
+		let mut chaining_key = [0u8; 32];
+		chaining_key.copy_from_slice(&chaining_key_vec);
+
+		let sending_key_vec = hex::decode("969ab31b4d288cedf6218839b27a3e2140827047f2c0f01bf5c04435d43511a9").unwrap();
+		let mut sending_key = [0u8; 32];
+		sending_key.copy_from_slice(&sending_key_vec);
+
+		let receiving_key_vec = hex::decode("bb9020b8965f4df047e07f955f3c4b88418984aadc5cdb35096b9ea8fa5c3442").unwrap();
+		let mut receiving_key = [0u8; 32];
+		receiving_key.copy_from_slice(&receiving_key_vec);
```

Comment:
Could you link to the test data in a comment for the test?

Reply:
Sure